### ISSUE TITLE
Parsing pub-data.json failed. Error: FormatException: Unexpected end of input (at character 1) 

### ISSUE THREAD

**Issue by sortie** [dart-lang/pub-dartlang-dart#1740]
The search logs are full of this error:

```
pub.search.backend: Parsing pub-data.json failed.

Error:
    FormatException: Unexpected end of input (at character 1)
    
    ^
    

Stack:
    #0      _ChunkedJsonParser.fail (dart:convert/runtime/libconvert_patch.dart:1358:5)
    #1      _ChunkedJsonParser.close (dart:convert/runtime/libconvert_patch.dart:511:7)
    #2      _parseJson (dart:convert/runtime/libconvert_patch.dart:30:10)
    #3      JsonDecoder.convert (dart:convert/json.dart:540:36)
    #4      JsonCodec.decode (dart:convert/json.dart:167:41)
    #5      SearchBackend._apiDocPagesFromPubDataText (package:pub_dartlang_org/search/backend.dart:151:29)
    #6      SearchBackend.loadDocuments (package:pub_dartlang_org/search/backend.dart:102:25)
    <asynchronous suspension>
    #7      BatchIndexUpdater._updateBatch (package:pub_dartlang_org/search/updater.dart:116:16)
    <asynchronous suspension>
    #8      BatchIndexUpdater.runTask (package:pub_dartlang_org/search/updater.dart:90:13)
    <asynchronous suspension>
    #9      TaskScheduler.run.runTask (package:pub_dartlang_org/shared/task_scheduler.dart:65:26)
    <asynchronous suspension>
    #10     TaskScheduler.run (package:pub_dartlang_org/shared/task_scheduler.dart:98:20)
    <asynchronous suspension>
    #11     _main.<anonymous closure> (file:///project/app/bin/service/search.dart:105:15)
    <asynchronous suspension>
    #12     withAppEngineServices.<anonymous closure> (package:appengine/src/appengine_internal.dart:46:51)
    #13     _withAppEngineServicesInternal.<anonymous closure> (package:appengine/src/appengine_internal.dart:115:20)
    <asynchronous suspension>
    #14     _ServiceScope._fork.<anonymous closure> (package:gcloud/service_scope.dart:210:19)
    #15     _rootRun (dart:async/zone.dart:1124:13)
    #16     _CustomZone.run (dart:async/zone.dart:1021:19)
    #17     _runZoned (dart:async/zone.dart:1516:10)
    #18     runZoned (dart:async/zone.dart:1463:12)
    #19     _ServiceScope._fork (package:gcloud/service_scope.dart:209:12)
    #20     fork (package:gcloud/service_scope.dart:104:30)
    #21     _withAppEngineServicesInternal (package:appengine/src/appengine_internal.dart:105:10)
    #22     withAppEngineServices (package:appengine/src/appengine_internal.dart:46:5)
    #23     withAppEngineServices (package:appengine/appengine.dart:108:10)
    #24     _main (file:///project/app/bin/service/search.dart:51:9)
    <asynchronous suspension>
    #25     _wrapper.<anonymous closure> (package:pub_dartlang_org/shared/service_utils.dart:226:22)
    <asynchronous suspension>
    #26     Chain.capture.<anonymous closure> (package:stack_trace/src/chain.dart:101:24)
    #27     _rootRun (dart:async/zone.dart:1124:13)
    #28     _CustomZone.run (dart:async/zone.dart:1021:19)
    #29     _runZoned (dart:async/zone.dart:1516:10)
    #30     runZoned (dart:async/zone.dart:1463:12)
    #31     Chain.capture (package:stack_trace/src/chain.dart:99:12)
    #32     _wrapper (package:pub_dartlang_org/shared/service_utils.dart:224:9)
    #33     _startIsolate.<anonymous closure> (dart:isolate/runtime/libisolate_patch.dart:292:17)
    #34     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:171:12)
```

Please fix or silence.

**Comment by isoos** [comment]
Empty `pub-data.json` files are not a good signs, we should look into it. Silencing the log warning because it doesn't add much info.

**Comment by isoos** [comment]
The additional logging revealed that the issue is coming from the package `googleapis`, it is rather large, and the dartdoc process throws an OOM exception:

````
Exhausted heap space, trying to allocate 32970864 bytes.
Unhandled exception:
Out of Memory
#0      PubDataGenerator.generate (package:pub_dartdoc/pub_data_generator.dart:73:5)
#1      _asyncErrorWrapperHelper.<anonymous closure> (dart:async/runtime/libasync_patch.dart:84:45)
#2      _RootZone.runBinary (dart:async/zone.dart:1384:54)
#3      _FutureListener.handleError (dart:async/future_impl.dart:143:20)
#4      Future._propagateToListeners.handleError (dart:async/future_impl.dart:654:47)
#5      Future._propagateToListeners (dart:async/future_impl.dart:675:24)
#6      Future._completeError (dart:async/future_impl.dart:494:5)
#7      Future._asyncCompleteError.<anonymous closure> (dart:async/future_impl.dart:542:7)
#8      _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#9      _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#10     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:115:13)
````

**Comment by isoos** [comment]
Local generation of the `pub-data.json` for `package:googleapis` produces a file about `32 198 kb` long. There are a few low-hanging fruits:
- building a tree upfront removes the parent references
- removing the URL and source references for items that we don't use anyway (e.g. class methods are interesting for their symbol and their documentation, but we always link to the class, not to a specific class method).

**Comment by isoos** [comment]
After #1895 the size is down to `20 703k` (-35.7% to the original size).

An experiment shows that with a further refactoring (removing the `parent` field and building a `children` list for each `parent`) if could go down to `16 438k`. However, the backward-compatible handling of that will be a bit more complex than it was with the `name`.

**Comment by isoos** [comment]
I've experimented, and transforming the `kind` key to groups e.g. `libraries`, `classes`, `methods` (respectively under their parent element) yields a bit more advantage.

**Comment by isoos** [comment]
The initial issue was fixed, and I think we are done with the low-effort changes. While we could do the refactor for hierarchical data, as long as it is not pressing for some reason, I'd just postpone it indefinitely.
